library/ssl_tls.c

```diff
@@ -7045,8 +7045,16 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
 }
 
 /*
- * Send application data to be encrypted by the SSL layer,
- * taking care of max fragment length and buffer size
+ * Send application data to be encrypted by the SSL layer, taking care of max
+ * fragment length and buffer size.
+ *
+ * According to RFC 5246 Section 6.2.1:
+ *
+ *      Zero-length fragments of Application data MAY be sent as they are
+ *      potentially useful as a traffic analysis countermeasure.
+ *
+ * Therefore, it is possible that the input message length is 0 and the
+ * corresponding return code is 0 on success.
```

Comment:
Our code is currently half-baked: `ssl_read` ignores a single application data record of size `0`, but if the next record fetched also is a length zero application record, this one is forwarded.

So we can neither claim to generally discard nor to generally forward AD records of length `0`.